lib/Doctrine/ORM/Tools/SchemaTool.php

```diff
@@ -138,6 +138,37 @@ private function processingNotRequired($class, array $processedClasses)
             ($class->isInheritanceTypeSingleTable() && $class->name !== $class->rootEntityName);
     }
 
+    /**
+     * Resolves fields in index mapping to column names
+     *
+     * @param ClassMetadata $class
+     * @param mixed[]       $indexData index or unique constraint data
+     *
+     * @return string[] Column names from combined fields and columns mappings
+     */
+    private function getIndexColumns($class, array $indexData)
+    {
+        $columns = [];
+
+        if (isset($indexData['columns'])) {
+            $columns = $indexData['columns'];
```

Comment:
I think its not a good idea to merge them, because in case of many the order of columns is significant for how a B-Tree index is created in the database. 

The only solution where we could support mixing them would be to have a `@IndexField` and `@IndexColumn` annotation and in xml nested children `<index-field name="foo"><index-column name="bar" />`. This is a bit overkill for this problem, so I would rather that we throw an exception if you define both instead of merging them.

What do you think?

Reply:
I have adapted code to enforce exactly one of fields/columns and extended tests to cover all cases I could come up with.